imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,485 @@
+        logger.warning(
+            "Frame epochs are not unique for each l1c product: "
+            f"{len(frame_epochs)} epochs vs {len(l1c_products)} products."
+        )
+
+    (combined_exptime_45, combined_exptime_90, combined_exptime_total) = (
+        build_dps_combined_exposure_time(l1c_products, spacing_deg)
+    )
+
+    energy_centers_from_file = l1c_products[0].energy_bin_center.values
+    energy_bin_edges, energy_midpoints = build_energy_bins()
+
+    # The function we call adds an extra bin edge at the start of the array.
+    # We remove it here.
+    energy_bin_edges = energy_bin_edges[1:]
+    energy_midpoints = energy_midpoints[1:]
+
+    if not np.allclose(energy_centers_from_file, energy_midpoints):
+        raise ValueError(
+            f"Energy bin centers from file: {energy_centers_from_file}\n"
+            f"Energy midpoints from function: {energy_midpoints}"
+        )
+
+    energy_delta = np.diff(energy_bin_edges)
+    if not np.all(energy_delta >= 0):
+        raise ValueError("Energy bin edges must be monotonically increasing.")
+
+    for prod_num, l1c_prod in enumerate(l1c_products):
+        time = float(l1c_prod.epoch.values)
+        logger.info(
+            f"Projecting exposure time from product at time index {prod_num} "
+            f"with epoch {time} = {spice.et2utc(time, 'ISOC', 3)}"
+        )
+
+        (
+            az_grid_raveled,
+            el_grid_raveled,
+            flat_indices_helio,
+            flat_indices_dps,
+            hae_az_in_dps_az,
+            hae_el_in_dps_el,
+            hae_az_in_dps_az_indices,
+            hae_el_in_dps_el_indices,
+        ) = project_inertial_frame_to_dps(
+            event_time=time,
+            existing_grids=(az_grid, el_grid),
+        )
+
+        # TODO: Replace with sum of counts across all l1c products on an inertial grid
+        pointing_projected_counts = l1c_prod.counts.values.reshape(
+            -1, l1c_prod.counts.shape[-1], order="F"
+        )[flat_indices_dps, :]
+
+        combined_counts += pointing_projected_counts
+
+    # TODO: this will be removed in the final version
+    # Output the last pointing's data for use in development and testing
+    developer_dict = {
+        "az_grid": az_grid,
+        "el_grid": el_grid,
+        "az_grid_raveled": az_grid_raveled,
+        "el_grid_raveled": el_grid_raveled,
+        "hae_az_in_dps_az": hae_az_in_dps_az,
+        "hae_el_in_dps_el": hae_el_in_dps_el,
+        "hae_az_in_dps_az_indices": hae_az_in_dps_az_indices,
+        "hae_el_in_dps_el_indices": hae_el_in_dps_el_indices,
+        "radii_helio": radii_helio,
+    }
+
+    return (
+        combined_counts,
+        combined_exptime_total,
+        combined_exptime_45,
+        combined_exptime_90,
+        frame_epochs,
+        energy_delta,
+        energy_bin_edges,
+        energy_midpoints,
+        solid_angle_grid,
+        developer_dict,
+    )
+
+
+def ultra_l2(l1c_products: list) -> xr.Dataset:
```

Comment:
If you look at examples of other instruments you will see that they all follow the same format for using this function:

def ultra_l1c(data_dict: dict, data_version: str) -> list[xr.Dataset]:

Reply:
I'm not convinced that the same approach will work for L2. It looks like generally, the CLI passes in a `data_dict[Logical_source] = dataset`. But when passing multiple L1c datasets, won't their Logical_source values be non-unique? I could have it pass them in as `f"{Logical_source}_{i}"` to avoid that, but I wasn't sure what using a dict adds over a list here. 

I'll switch to a dict for consistency regardless.